File: 4th/doublelinked.py
class Node:
    def __init__(self, data):
        self.data=data
        self.prev=None
        self.next=None

class DoubleLinkedList:
    def __init__(self):
        self.head=None
        self.tail=None
    
    def append(self, data):
        new_node=Node(data)
        if self.head is None:
            self.head=new_node
            self.tail=new_node
        else:
            new_node.prev=self.tail
            self.tail.next=new_node
            self.tail=new_node

    def printlist(self):
        current=self.head
        while current:
            print(current.data, end=' -> ')
            current=current.next
        print("null")

if __name__ == "__main__":
    myList=DoubleLinkedList()
    print("첫번째 숫자를 입력하세요")
    a=input()
    print("두번째 숫자를 입력하세요")
    b=input()
    print("세번째 숫자를 입력하세요")
    c=input()
    myList.append(a)
    myList.append(b)
    myList.append(c)

    myList.printlist()

File: 4th/doublelinked.c
#include <stdio.h>
#include <stdlib.h>
#define _CRT_SECURE_NO_WARNINGS

//노드 structure 정의
struct Node {
	int data;
	struct Node* next;
	struct Node* prev;
};//더블 링크드 리스트이므로 처음과 끝 모두 정의
//리스트 정의
struct DoubleLinkedList {
	struct Node* head;
	struct Node* tail;
};
//초기화
void init(struct DoubleLinkedList* list) {
	list->head = NULL;
	list->tail = NULL;
}
//노드 생성 코드
struct Node* createNode(int data) {
	struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
	newNode->data = data;
	newNode->prev = NULL;
	newNode->next = NULL;
}
//노드 추가 코드
void append(struct DoubleLinkedList* list, int data) {
	struct Node* newNode = createNode(data);
	if (list->tail == NULL) {//만약 노드가 처음 상태라면
		list->head = newNode;//처음과 끝 모두 자기자신으로 설정
		list->tail = newNode;
	}
	else {
		list->tail->next = newNode;
		newNode->prev = list->tail;
		list->tail = newNode;
	}
}
//리스트 출력 코드
void printlist(struct DoubleLinkedList* list) {
	struct Node* current = list->head;
	while (current != NULL) {
		printf("%d -> ", current->data);
		current = current->next;
	}
	printf("NULL\n");
}

int main() {
	struct DoubleLinkedList myList;
	init(&myList);
	int a, b, c;
	printf("첫번째 노드 값: ");
	scanf("%d", &a);
	printf("두번째 노드 값: ");
	scanf("%d", &b);
	printf("세번째 노드 값: ");
	scanf("%d", &c);
	append(&myList, a);
	append(&myList, b);
	append(&myList, c);
	printlist(&myList);
	return 0;
}